Simplify enums using .allCases from Swift 4.2

ghowen opened the pull request:
PR in regards to https://github.com/LiveUI/Awesome/issues/21

- Simplified enums by removing as many stored properties as possible and replacing them with computed ones. For this mainly the generator and the Awesome.xxx extensions needed to be changed.
- Some renaming to be more Swift like but keeping downward compatibility through deprecation warnings. The enums are now capitalized, i.e. FontAwesome.Solid.car instead of FontAwesome.solid.car.
- updated enums to FontAwesome 5.4.2

I am not so familiar with creating pull requests so please let me know if this is OK like this. Thanks.

rafiki270 commented:
Looks good but do you think you could do descriptions for all public variables and methods there for autocomplete? If we’re to upgrade a major version then we should improve all of the interface I think

divadretlaw commented on Classes/Extensions/Amazing+Tools.swift:
```diff
@@ -26,181 +25,317 @@ public extension Amazing {
+    }
+    
+    public var description: String {
+        return String(describing: self).split(separator: "-").map { $0.firstUppercased }.joined(separator: " ")
```
What are you trying to achieve with this actually? There are no hyphens in the keys therefore this does essentially nothing.

`Awesome.Solid.americanSignLanguageInterpreting.label` printed "American Sign Language Interpreting" before and now it prints "AmericanSignLanguageInterpreting" therefore `description` lost value compared to before.

Same thing for the key. Before it contained hyphens now it does not.

Not sure how to store the descriptions and keys.

But here are several possibilities

* Store them in an array as before
* Store them in a dictionary
* Generate a `switch` which returns the specific label / name (description / key)
    - `-` Makes the files very long
* Make the RawValue a custom type which stores all needed values (unicode, name and label)
    - `+` Shorter enum files
    - `-` Makes the RawValue inits slightly more expensive. And there are a lot

ghowen replied:
Sorry, my bad. The idea was to keep backward compatibility but it's obviously not working like this. I still think it's possible to do it with a computed property, converting the camel caps with regex i.e. Will fix it and submit.

divadretlaw commented on Generator/Sources/Generator/EnumBuilder.swift:
```diff
@@ -12,7 +12,9 @@ func buildEnum (_ name: String, from styles: [String: [Icon]]) -> String {
-        content += "\n    public enum \(style): String, Amazing {\n"
+        content += "\n    @available(*, deprecated, renamed: \"\(style.firstUppercased())\")"
+        content += "\n    public typealias \(style) = \(style.firstUppercased())"
+        content += "\n    public enum \(style.firstUppercased()): String, CaseIterable, Amazing {\n"
```
The `CaseIterable` should be move to `Amazing`

ghowen commented on Generator/Sources/Generator/EnumBuilder.swift:
```diff
@@ -12,7 +12,9 @@ func buildEnum (_ name: String, from styles: [String: [Icon]]) -> String {
 
-        content += "\n    public enum \(style): String, Amazing {\n"
+        content += "\n    @available(*, deprecated, renamed: \"\(style.firstUppercased())\")"
+        content += "\n    public typealias \(style) = \(style.firstUppercased())"
```
Brilliant. I tried to do that before but failed as the computed properties reference `self`. Using generics solved just that. Will change as suggested.

rafiki270 commented:
If the interface changes then we up a major version so youguys don’t have to worry about compatibility

ghowen commented:
OK, I implemented the changes suggested by @divadretlaw and also solved the bug with the keys with a computed property. `Amazing` cannot implement `CustomStringConvertible` though as this triggers an endless loop due to reflection in the computed `description` property. But that's no big deal. 

Being not so experienced with GitHub and creating PRs I did not realize that @divadretlaw already made some code changes that I could have just accepted. So I did the changes myself in my fork which lead to the suggested changes being outdated. Sorry for this, next time.

@rafiki270 The interface is compatible but with deprecation notices. I'll leave versioning up to you guys.

divadretlaw commented:
Did some cleanup as the `@available` stuff in the protocol did not translate into actual deprecations. Upped the Swift version in all targets and some other general cleanup.

rafiki270 commented:
Could we please comment on all the public vars and functions if we are raising the new version?

rafiki270 commented:
Happy to merge right after

ghowen commented:
@rafiki270 Not fully sure what you mean. Can you give an example of what you have in mind? It looks like autocomplete comments have been removed from Xcode some time ago: 
https://stackoverflow.com/questions/41830473/xcode-8-2-1-not-showing-documentation-description-on-autocomplete/43982094#43982094

rafiki270 commented:
@ghowen https://developer.apple.com/library/archive/documentation/Xcode/Reference/xcode_markup_formatting_ref/index.html basic is `///`

rafiki270 commented:
you can also go before the property or function and do `cmd+/`

ghowen commented:
@rafiki270 Gotcha. I was missing the `///`. `cmd+/` creates only `//` on my machine and that did not work as discussed on SO.

rafiki270 commented:
sorry, `cmd+option+/` ... it will generate even input/output properties on methods, etc ...

ghowen commented:
@rafiki270 OK, I hope this does what you meant. I also added another computed property to Amazing and removed `CaseIterable` from the enums as they get this through `Amazing`. 

BTW: The shortcut to document does not work (at least on some) non US keyboards. But I found it in the menu and changed it locally. Very helpful!

rafiki270 commented on Classes/Extensions/Amazing+Tools.swift:
```diff
@@ -9,198 +9,184 @@
-        return name
+extension Amazing {
+    
+    public var detailedKey: String {
```
These still missing comments on what they do?

ghowen replied:
They were already in declaration of the `Amazing` protocol and I thought this would be enough. But it looks like the properties appear twice in autocomplete: once with (protocol) and once without (extension) the hint. I'll add them to the extension as well.

ghowen replied:
Looks like the same thing applies for the deprecation warnings. So I'll add them to the protocol again too.

rafiki270 commented on Classes/Extensions/Amazing+Tools.swift:
```diff
@@ -9,198 +9,190 @@
-        let name = AwesomeType.keys[index]
-        return name
+    
+    @available(swift, deprecated: 4.2, message: "This will be removed in the future. Please use .allKeys instead")
```
You can also do `@available(*, unavailable, renamed: "allKeys")` which will offer user the "red dot of fix" in the code ... this is how apple generates their "Fix it!" warnings ...

rafiki270 replied:
I would not worry about compatibility as this will be a major update `2.0.0`

ghowen replied:
That's a bit problematic as the computed property expects a return type that `fatalError` cannot meet. Addind `fatalError` before the `return` with the correct type generates compiler warnings that the return will never be executed. IMHO there is no need for the fatalError as the compiler will fail on the `unavailable` anyway. I'll have another look and submit a proposal.

rafiki270 replied:
have you tried? :) ... you need to put it in the `get` method, set can be let empty ...

```swift
var prop: String {
   get {
      fatalError()
   }
   set { }
}
```

ghowen replied:
Nice! I it works when using get/set but not when the properties are declared as read-only computed properties.

rafiki270 commented:
Looks good to me ... @ghowen would you like to merge and do a release? It will be `2.0.0` and please describe the changes made in the release notes :) ... good job man!

ghowen commented:
@rafiki270 Thanks. I've never done that on a public repo and honestly I don't know what else needs to be changed for the release to work properly (like version numbers, pod configs, carthage configs etc.). So maybe someone more experienced can do this one and I'll have a close look at all the changes that were needed for the rollout. Sound OK?

rafiki270 commented:
@ghowen don't forget to add yourself to the list of contributors on README page! :)

ghowen commented:
:) Sure, thanks. What do you need to do to update the pod and carthage lib? I saw that the pod is still on 1.3.0 while the lib was already on 1.4.0.

rafiki270 commented:
Also, should you have a read only prop then skip the setter (re fatal error in vars and lets)
